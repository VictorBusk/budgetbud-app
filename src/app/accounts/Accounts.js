import React, {PureComponent} from 'react';
import {FlatList} from 'react-native';
import {connect} from 'react-redux';
import _ from 'lodash';
import {
    AppHeader,
    Separator
} from "../../components/index";
import {
    Container,
    ListItem,
    Body,
    Right,
    Text,
    Spinner,
    CheckBox,
    Button
} from 'native-base';
import I18n from "../../strings/i18n";
import {
    button,
    text,
    list,
    container, color
} from "../../style/index";
import {showWarningToast} from "../../helpers/index";
import {
    accountsSelected,
    getAccounts,
    linkAccounts,
    resetAccountsError
} from "../../redux/actions";

class Accounts extends PureComponent {
    componentWillMount() {
        if (!this.props.accountsInitialized)
            this.props.getAccounts();
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.accountsError) {
            showWarningToast(nextProps.accountsError);
            this.props.resetAccountsError();
        }
    }

    onSavePress = () => {
        this.props.linkAccounts(this.props.linkedAccounts, () => {
            this.props.navigation.pop();
        });
    };

    onCheckBoxPress = ({accountID}) => {
        let tmp = this.props.linkedAccounts;

        if (tmp.includes(accountID)) {
            tmp.splice(tmp.indexOf(accountID), 1)
        } else {
            tmp.push(accountID);
        }

        this.props.accountsSelected(tmp);
    };

    render() {
        return (
            <Container style={container.signedInContainer}>
                <AppHeader headerText={I18n.t('accountsHeader')}
                           showBackButton={true}
                           onLeftButtonPress={() => {
                               if (!this.props.linkLoading)
                                   this.props.navigation.pop()
                           }}
                />

                <Container style={{justifyContent: 'center'}}>
                    {this.props.accountsLoading ? (
                        <Spinner style={{
                            alignItems: 'center',
                            justifyContent: 'center'
                        }} color='#1c313a'/>) : (
                        <FlatList
                            data={this.props.accountItems}
                            renderItem={this.renderItem}
                        />
                    )}
                </Container>

                <Separator/>

                <Button rounded
                        onPress={() => {
                            if (!this.props.linkLoading || !this.props.accountsLoading) {
                                this.onSavePress()
                            }
                        }}
                        style={[button.defaultButton, color.button]}
                >
                    {this.props.linkLoading ? (
                        <Spinner color='#D0D0D0'/>) : (
                        <Text style={text.submitButtonText}>{I18n.t('accountsSaveButton')}</Text>
                    )}
                </Button>
            </Container>
        );
    }

    renderItem = ({item}) => {
        const checked = this.props.linkedAccounts.includes(item.accountID);

        return (
            <ListItem>
                <Body>
                <Text style={color.text}>{item.name}</Text>
                </Body>
                <Right style={list.listItemCheckBoxPadding}>
                    <CheckBox
                        style={checked ? color.checkboxChecked : color.checkboxUnchecked}
                        checked={checked}
                        onPress={() => this.onCheckBoxPress(item)}
                    />
                </Right>
            </ListItem>
        );
    }
}

const mapStateToProps = ({account}) => {
    const {
        accounts,
        linkedAccounts,
        accountsLoading,
        linkLoading,
        accountsError,
        accountsInitialized
    } = account;

    const accountItems = _.map(accounts, (item, key) => {
        return {name: item.name, accountID: item.id, key: key};
    });

    return {
        accountItems,
        linkedAccounts,
        accountsLoading,
        linkLoading,
        accountsError,
        accountsInitialized
    };
};

const mapDispatchToProps = {
    getAccounts,
    linkAccounts,
    accountsSelected,
    resetAccountsError
};

export default connect(mapStateToProps, mapDispatchToProps)(Accounts);
