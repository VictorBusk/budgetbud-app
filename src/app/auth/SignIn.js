import React, {Component} from 'react';
import {
    KeyboardAvoidingView,
    Keyboard,
    TouchableWithoutFeedback, View
} from 'react-native';
import {connect} from 'react-redux';
import {Container, Button, Label} from 'native-base';
import {AuthForm, Logo} from '../../components/index';
import {container, button, color} from "../../style/index";
import I18n from "../../strings/i18n";
import {
    codeChanged,
    cprNumberChanged,
    resetAuthState,
    signIn,
    resetAuthError
} from "../../redux/actions";
import {showWarningToast} from "../../helpers/toasts";

class SignIn extends Component {
    componentWillReceiveProps(nextProps) {
        if (nextProps.authError) {
            showWarningToast(nextProps.authError);
            this.props.resetAuthError();
        }
    }

    onCprNumberChange = (text) => {
        this.props.cprNumberChanged(text);
    };

    onCodeChange = (text) => {
        this.props.codeChanged(text);
    };

    handleSubmit = () => {
        Keyboard.dismiss();
        const {cprNumber, code} = this.props;
        this.props.signIn({cprNumber, code});
    };

    onForgotCodeButtonPress = () => {
        if (!this.props.authLoading) {
            this.props.resetAuthState(() => {
                this.props.navigation.navigate('RequestActivationCode');
            });
        }
    };

    onGoToSignUpButtonPress = () => {
        if (!this.props.authLoading) {
            this.props.resetAuthState(() => {
                this.props.navigation.navigate('SignUp');
            });
        }
    };

    render() {
        return (
            <KeyboardAvoidingView behavior='padding' style={container.signedOutContainer}>
                <TouchableWithoutFeedback onPress={Keyboard.dismiss} accessible={false}>
                    <Container style={{alignSelf: 'stretch'}}>

                        <View style={[container.parentContainer, {flex: 2}]}>
                            <Logo logoText={I18n.t('signInLogoWelcome')}/>
                        </View>

                        <AuthForm handleSubmit={this.handleSubmit}
                                  onCprNumberChange={this.onCprNumberChange}
                                  onSecondInputChange={this.onCodeChange}
                                  cprNumber={this.props.cprNumber}
                                  secondInput={this.props.code}
                                  authLoading={this.props.authLoading}
                                  isSignIn={true}
                        />

                        <Container>
                            <Container style={container.optionContainer}>
                                <Button transparent style={button.optionButton}
                                        onPress={() => this.onForgotCodeButtonPress()}>
                                    <Label style={color.optionButton}>{I18n.t('signInForgotCode')}</Label>
                                </Button>

                                <Button transparent style={button.optionButton}
                                        onPress={() => this.onGoToSignUpButtonPress()}>
                                    <Label style={color.optionButton}>{I18n.t('signInGoToSignUp')}</Label>
                                </Button>
                            </Container>
                        </Container>

                    </Container>
                </TouchableWithoutFeedback>
            </KeyboardAvoidingView>
        );
    };
}

const mapStateToProps = ({auth}) => {
    return {cprNumber, code, authError, authLoading} = auth;
};

const mapDispatchToProps = {
    signIn,
    cprNumberChanged,
    codeChanged,
    resetAuthState,
    resetAuthError
};

export default connect(mapStateToProps, mapDispatchToProps)(SignIn);
