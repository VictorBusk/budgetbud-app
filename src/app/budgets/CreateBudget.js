import React, {Component} from 'react';
import {Keyboard} from 'react-native';
import {Container} from 'native-base';
import {connect} from 'react-redux';
import {BudgetForm, AppHeader} from "../../components";
import I18n from "../../strings/i18n";
import {container} from "../../style";
import {
    showWarningToast,
    checkInputAmount,
    correctConversion,
    setupNewCategoriesList
} from "../../helpers";
import {
    createBudget,
    getLinkedAccounts,
    createCategories,
    mapExpensesToBudget
} from "../../redux/actions";
import {resetBudgetError} from "./budget_actions";
import {resetCategoriesError} from "../categories/category_actions";

class CreateBudget extends Component {
    state = {
        tmpIncome: this.props.income,
        tmpDisposable: this.props.disposable,
        tmpTotalGoalsAmount: this.props.totalGoalsAmount,
        tmpCategories: [],
        submitLoading: false
    };

    async componentWillMount() {
        this.setState({
            tmpCategories: await this.props.mapExpensesToBudget()
        })
    };

    componentWillReceiveProps(nextProps) {
        if (nextProps.budgetError) {
            showWarningToast(nextProps.budgetError);
            this.props.resetBudgetError();
        } else if (nextProps.categoriesError) {
            showWarningToast(nextProps.categoriesError);
            this.props.resetCategoriesError();
        }
    }

    onIncomeChange = (newIncome) => {
        newIncome = correctConversion(newIncome);
        if (checkInputAmount(newIncome)) {
            const incomeDiff = newIncome - this.state.tmpIncome;
            const newDisposable = Math.round((this.state.tmpDisposable + incomeDiff) * 100) / 100;

            this.setState({
                tmpIncome: newIncome,
                tmpDisposable: newDisposable,
            })
        }
    };

    onCategoryChange = (name, oldAmount, newAmount) => {
        newAmount = correctConversion(newAmount);
        if (checkInputAmount(newAmount)) {
            const categoryDiff = oldAmount - newAmount;
            const newDisposable = Math.round((this.state.tmpDisposable + categoryDiff) * 100) / 100;
            const newTotalGoalsAmount = Math.round((this.state.tmpTotalGoalsAmount - categoryDiff) * 100) / 100;
            const newTmpCategories = setupNewCategoriesList(
                this.state.tmpCategories,
                name,
                newAmount);

            this.setState({
                tmpDisposable: newDisposable,
                tmpTotalGoalsAmount: newTotalGoalsAmount,
                tmpCategories: newTmpCategories
            })
        }
    };

    handleSubmit = async () => {
        Keyboard.dismiss();
        this.setState({
            submitLoading: true
        });

        const categoryItems = this.state.tmpCategories.filter((obj) => {
            return obj.amount > 0
        });

        await this.props.createBudget(
            this.state.tmpIncome,
            this.state.tmpDisposable,
            this.state.tmpTotalGoalsAmount,
            (budgetID) => {
                this.props.createCategories
                (
                    budgetID,
                    categoryItems,
                    () => {
                        this.setState({
                            submitLoading: false
                        });
                    }
                )
            }
        );
    };

    render() {
        return (
            <Container style={[container.signedInContainer, {alignItems: 'stretch'}]}>
                <AppHeader headerText={I18n.t('createBudgetHeader')}
                           infoText={I18n.t('createBudgetInfo')}
                           showBackButton={true}
                           onLeftButtonPress={() => {
                               if (!this.state.submitLoading)
                                   this.props.navigation.pop()
                           }}/>

                <BudgetForm handleSubmit={this.handleSubmit}
                            onIncomeChanged={this.onIncomeChange}
                            onCategoryChanged={this.onCategoryChange}
                            budgetID={this.props.budgetID}
                            tmpIncome={this.state.tmpIncome}
                            tmpTotalGoalsAmount={this.state.tmpTotalGoalsAmount}
                            tmpDisposable={this.state.tmpDisposable}
                            tmpCategories={this.state.tmpCategories}
                            debts={this.props.debts}
                            budgetLoading={this.props.budgetLoading}
                            submitLoading={this.state.submitLoading}
                            categoriesLoading={this.props.categoriesLoading}
                            budgetError={this.props.budgetError}
                />
            </Container>
        );
    }
}

const mapStateToProps = (state) => {
    const {
        income,
        debts,
        budgetLoading,
        budgetError,
        budgetID
    } = state.budget;

    const disposable = state.disposable.disposable;

    const linkedAccounts = state.account.linkedAccounts;
    const {
        categoriesLoading,
        categoriesError,
        totalGoalsAmount
    } = state.category;

    return {
        income,
        budgetID,
        linkedAccounts,
        debts,
        totalGoalsAmount,
        disposable,
        budgetLoading,
        categoriesError,
        budgetError,
        categoriesLoading
    };
};

const mapDispatchToProps = {
    createBudget,
    getLinkedAccounts,
    mapExpensesToBudget,
    createCategories,
    resetBudgetError,
    resetCategoriesError
};

export default connect(mapStateToProps, mapDispatchToProps)(CreateBudget);
