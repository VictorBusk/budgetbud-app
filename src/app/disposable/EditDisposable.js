import React, {Component} from 'react';
import {Container} from 'native-base';
import {connect} from "react-redux";
import I18n from "../../strings/i18n";
import {AppHeader, DisposableForm} from "../../components";
import {container} from "../../style";
import {showWarningToast} from "../../helpers";
import {checkInputAmount, correctConversion} from "../../helpers";
import {
    calculateDisposableCategoryDifferences,
    disposableChanged,
    resetDisposableError,
    resetDisposableForm,
    setTmpDisposable,
    categoriesSelected,
    getCategories
} from "../../redux/actions";

class EditDisposable extends Component {
    componentWillMount() {
        this.props.setTmpDisposable();
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.disposableError) {
            showWarningToast(nextProps.disposableError);
            this.props.resetDisposableError();
        }
    }

    onDisposableChange = (newDisposable) => {
        newDisposable = correctConversion(newDisposable);
        if (checkInputAmount(newDisposable)) {
            this.props.disposableChanged(newDisposable);
        }
    };

    onCheckBoxPress = ({categoryID}) => {
        let tmp = this.props.selectedCategories;

        if (tmp.includes(categoryID)) {
            tmp.splice(tmp.indexOf(categoryID), 1)
        } else {
            tmp.push(categoryID);
        }

        this.props.categoriesSelected(tmp);
    };

    onContinuePress = async () => {
        if (this.props.disposableCalculationLoading)
            return;

        await this.props.calculateDisposableCategoryDifferences(
            this.props.disposable,
            this.props.tmpDisposable,
            this.props.selectedCategories, () => {
                this.props.navigation.navigate('DisposablePreview');
            });
    };

    render() {
        return (
            <Container style={container.signedInContainer}>
                <AppHeader headerText={I18n.t('editDisposableHeader')}
                           showBackButton={true}
                           onLeftButtonPress={() => this.props.navigation.pop()}/>

                <DisposableForm disposableChanged={this.props.disposableChanged}
                                categoriesSelected={this.props.categoriesSelected}
                                categoryItems={this.props.categoryItems}
                                selectedCategories={this.props.selectedCategories}
                                budgetLoading={this.props.budgetLoading}
                                categoriesLoading={this.props.categoriesLoading}
                                disposable={this.props.disposable}
                                tmpDisposable={this.props.tmpDisposable}
                                disposableError={this.props.disposableError}
                                disposableCalculationLoading={this.props.disposableCalculationLoading}
                                disposableLoading={this.props.disposableLoading}
                                subtractionsLoading={this.props.subtractionsLoading}
                                onDisposableChanged={this.onDisposableChange}
                                onCheckBoxPressed={this.onCheckBoxPress}
                                onContinuePressed={this.onContinuePress}/>
            </Container>
        );
    }
}

const mapStateToProps = (state) => {
    const budgetLoading = state.budget.budgetLoading;
    const {
        disposable,
        tmpDisposable,
        disposableError,
        disposableLoading,
        disposableCalculationLoading
    } = state.disposable;
    const {
        categories,
        selectedCategories,
        categoriesLoading,
        subtractionsLoading,
        categoriesError
    } = state.category;

    const categoryItems = categories.filter((obj) => {
        return obj.amount > 0
    });

    return {
        budgetLoading,
        selectedCategories,
        categoryItems,
        categoriesLoading,
        subtractionsLoading,
        categoriesError,
        disposable,
        tmpDisposable,
        disposableError,
        disposableLoading,
        disposableCalculationLoading
    };
};

const mapDispatchToProps = {
    disposableChanged,
    getCategories,
    resetDisposableForm,
    calculateDisposableCategoryDifferences,
    categoriesSelected,
    setTmpDisposable,
    resetDisposableError
};

export default connect(mapStateToProps, mapDispatchToProps)(EditDisposable);
