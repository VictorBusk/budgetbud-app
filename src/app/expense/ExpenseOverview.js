import React, {Component} from 'react';
import {
    Body,
    Container,
    Icon,
    ListItem,
    Text,
    View,
    Right,
    Spinner
} from "native-base";
import {AppHeader, Separator} from "../../components";
import {FlatList, StyleSheet, TouchableOpacity} from "react-native";
import {connect} from "react-redux";
import I18n from "../../strings/i18n";
import {color, container} from "../../style";
import {
    getCategoryAlarms,
    toggleCategoryAlarm,
    getExpensesOfMonth
} from "../../redux/actions";

class ExpenseOverview extends Component {
    async componentWillMount() {
        if (!this.props.categoryAlarmsInitialized)
            this.props.getCategoryAlarms(this.props.budgetID);

        if (!this.props.expensesInitialized)
            this.props.getExpensesOfMonth();
    }

    onAlarmPress = async ({categoryID}) => {
        if (!this.props.toggleLoading)
            this.props.toggleCategoryAlarm(categoryID, this.props.budgetID)
    };

    render() {
        return (
            <Container style={container.signedInContainer}>
                <AppHeader headerText={I18n.t('expenseOverviewHeader')}
                           onLeftButtonPress={() => this.props.navigation.navigate("DrawerOpen")}/>
                <Container>
                    {this.props.expensesLoading ? (
                        <Spinner style={{
                            flex: 1
                        }} color='#1c313a'/>) : (
                        <Container>
                            <View style={{flex: 0.82}}>
                                <FlatList
                                    data={this.props.categoryItems}
                                    renderItem={this.renderCategory}
                                    keyExtractor={item => item.categoryID}
                                />
                                <Separator/>
                            </View>
                            <View style={{flex: 0.18}}>
                                <View style={[styles.budgetSummary, {paddingVertical: '5%',}]}>
                                    <View style={[styles.incomeFormStyle, {flex: 1}]}>
                                        <Text style={styles.listText}>
                                            {I18n.t('expenseOverviewTotalDebtsPerMonth')}
                                        </Text>
                                        <Text style={styles.listText}>
                                            {this.props.totalDebtPerMonth} {I18n.t('currency')}
                                        </Text>
                                    </View>
                                    <View style={[styles.incomeFormStyle, {flex: 1}]}>
                                        <Text style={styles.listText}>
                                            {I18n.t('expenseOverviewTotalExpenses')}
                                        </Text>
                                        <Text style={styles.listText}>
                                            {this.props.totalExpenses} {I18n.t('currency')}
                                        </Text>
                                    </View>
                                    <View style={[styles.incomeFormStyle, {flex: 1}]}>
                                        <Text style={styles.listText}>
                                            {I18n.t('expenseOverviewDisposable')}
                                        </Text>
                                        <Text style={styles.listText}>
                                            {this.props.disposable} {I18n.t('currency')}
                                        </Text>
                                    </View>
                                </View>
                            </View>
                        </Container>
                    )}
                </Container>
            </Container>
        );
    }

    renderCategory = ({item}) => {
        const notificationsEnabled = this.props.categoryAlarms.includes(item.categoryID);

        return (
            <ListItem>
                <View style={{width: '100%', paddingHorizontal: 4}}>
                    <View style={[container.removeIndenting, {flexDirection: 'row'}]}>
                        <Body style={{alignItems: 'flex-start'}}>
                        <Text style={color.text}>{item.name}</Text>
                        </Body>
                        <Right>
                            <TouchableOpacity onPress={() => this.onAlarmPress(item)}>
                                <Icon style={[color.darkIcon, {fontSize: 26}]}
                                      name={notificationsEnabled ? "ios-notifications" : "ios-notifications-outline"}/>
                            </TouchableOpacity>
                        </Right>
                    </View>
                    <View style={[container.removeIndenting, {flexDirection: 'row'}]}>
                        <Body style={{alignItems: 'flex-start'}}>
                        <Text style={{color: '#808080'}}>{I18n.t('expenseOverviewBudget')}</Text>
                        <Text style={{color: '#808080'}}>{I18n.t('expenseOverviewExpenses')}</Text>
                        <Text style={{color: '#808080'}}>{I18n.t('expenseOverviewRemaining')}</Text>
                        </Body>
                        <Right>
                            <Text style={{color: '#808080'}}>{item.budget} {I18n.t('currency')}</Text>
                            <Text style={{color: '#808080'}}>{item.expenses} {I18n.t('currency')}</Text>
                            <Text
                                style={((item.budget - item.expenses) < 0 ? {color: '#EB7C83'} :
                                    {color: '#808080'})}>{item.budget - item.expenses} {I18n.t('currency')}
                            </Text>
                        </Right>
                    </View>
                </View>
            </ListItem>
        );
    };
}

const styles = StyleSheet.create({
    buttonStyle: {
        justifyContent: 'flex-end',
        alignSelf: 'flex-end',
        paddingBottom: 10
    },
    incomeFormStyle: {
        alignSelf: 'center',
        width: '100%',
        paddingRight: 18,
        paddingLeft: 15,
        justifyContent: 'space-between',
        flexDirection: 'row'
    },
    expenseListStyle: {
        alignSelf: 'center',
        width: '90%',
        justifyContent: 'space-between',
        flexDirection: 'row',
        paddingTop: 23
    },
    itemStyle: {
        fontWeight: '600',
        alignSelf: 'flex-start',
        color: 'white'
    },
    spacedText: {
        flexGrow: 1,
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center',
        marginHorizontal: '5%',
        marginVertical: 10
    },
    budgetSummary: {
        marginBottom: 0,
        justifyContent: 'space-between',
        flex: 1
    },
    textStyle: {
        fontWeight: '400',
        fontSize: 14,
        alignSelf: 'flex-start',
        marginLeft: 5,
        color: '#00263A'
    },
    listText: {
        marginLeft: 5,
        alignSelf: 'flex-start',
        fontSize: 16,
        color: '#00263A'
    },
    modalButton: {
        flexDirection: 'column',
        justifyContent: 'center',
        alignItems: 'center'
    }
});

const mapStateToProps = (state) => {
    const {budgetID} = state.budget;
    const {debts} = state.debt;
    const {expenses, totalExpenses, expensesLoading, expensesInitialized} = state.expense;
    const disposable = state.disposable.disposable;
    const {categories} = state.category;
    const {categoryAlarms, toggleLoading, categoryAlarmsInitialized} = state.alarm;

    let totalDebtPerMonth = 0;
    debts.forEach(d => totalDebtPerMonth += d.debtData.amountPerMonth);

    const categoryItems = [];

    categories.forEach(category => {
        const index = expenses.findIndex(x => x.categoryTypeID === category.categoryTypeID);

        if (index !== -1) {
            categoryItems.push({
                categoryID: category.categoryID,
                name: category.name,
                budget: category.amount,
                expenses: expenses[index].amount,
                notificationsEnabled: false
            });
        }
    });

    return {
        budgetID,
        categoryAlarms,
        toggleLoading,
        totalDebtPerMonth,
        disposable,
        expenses,
        totalExpenses,
        expensesLoading,
        categoryItems,
        expensesInitialized,
        categoryAlarmsInitialized
    }
};

export default connect(mapStateToProps, {
    getExpensesOfMonth, getCategoryAlarms, toggleCategoryAlarm
})(ExpenseOverview);
