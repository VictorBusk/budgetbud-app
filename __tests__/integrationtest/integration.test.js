import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {INITIAL_ACCOUNT_STATE, INITIAL_AUTH_STATE, INITIAL_BUDGET_STATE} from "../test_helper/initial_state";
import {
    CREATE_BUDGET,
    CREATE_BUDGET_SUCCESS,
    EDIT_DISPOSABLE,
    EDIT_DISPOSABLE_SUCCESS,
    GET_EXPENSES_OF_MONTH,
    GET_EXPENSES_OF_MONTH_SUCCESS,
    GET_LINKED_ACCOUNTS,
    GET_LINKED_ACCOUNTS_SUCCESS,
    SIGN_IN
} from "../../src/strings/types";
import firebase from 'firebase';
import axios from "axios";
import store from '../../src/redux/index';
import authReducer from '../../src/app/auth/auth_reducer';
import accountReducer from '../../src/app/accounts/account_reducer';
import CreateBudget from "../../src/app/budgets/CreateBudget";
import SignIn from "../../src/app/auth/SignIn";
import Accounts from "../../src/app/accounts/Accounts";
import {setupFirebaseMock} from "../test_helper/firebase_mock";
import ExpenseOverview from "../../src/app/expenses/ExpenseOverview";
import EditDisposable from "../../src/app/disposable/EditDisposable";

const authActions = require('../../src/app/auth/auth_actions');
const budgetActions = require('../../src/app/budgets/budget_actions');
const accountActions = require('../../src/app/accounts/account_actions');
const expenseActions = require('../../src/app/expenses/expense_actions');
const disposableActions = require('../../src/app/disposable/disposable_actions');

Enzyme.configure({adapter: new Adapter()});
describe('SignIn', () => {
    it('renders SignIn, envoke sign in action, update state and rerender SignIn', async () => {
        // Setup initial screen
        const initialWrapper = shallow(
            <SignIn/>,
            {context: {store}}
        );
        expect(initialWrapper.dive()).toMatchSnapshot();

        const signInWithCustomToken = jest.fn();
        jest.spyOn(firebase, 'auth').mockImplementation(() => {
            return {signInWithCustomToken}
        });

        // Fire action
        const testToken = '1234';
        const cprNumber = '1234567890';
        const code = '1234';

        const expectedAction = [
            {type: SIGN_IN}
        ];
        const resp = {data: {token: testToken}};
        axios.post.mockResolvedValue(resp);
        let newState = INITIAL_AUTH_STATE;

        store.dispatch(await authActions.signIn({cprNumber, code})).then(() => {
            newState = authReducer(newState, expectedAction[0]);
            expect(store.getState().auth).toEqual(newState)
            expect(firebase.auth).toHaveBeenCalledTimes(1);
            expect(signInWithCustomToken).toHaveBeenCalledWith(testToken);
        });

        const afterWrapper = shallow(
            <SignIn/>,
            {context: {store}},
        );
        expect(afterWrapper.dive()).toMatchSnapshot();
    });
});

describe('CreateBudget', () => {
    beforeEach(() => {
        setupFirebaseMock();
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    it('renders CreateBudget, envoke sign in action, update state and rerender CreateBudget', async () => {
        // Setup initial screen
        const initialWrapper = shallow(
            <CreateBudget/>,
            {context: {store}}
        );
        expect(initialWrapper.dive()).toMatchSnapshot();

        const tmpIncome = 25;
        const tmpDisposable = 50;
        const tmpTotalGoalsAmount = 80;
        const postResult = {data: {id: 123}};

        const postMock = axios.post;
        postMock.mockImplementation(() =>
            Promise.resolve(postResult)
        );

        const expectedAction = [
            {type: CREATE_BUDGET},
            {
                type: CREATE_BUDGET_SUCCESS,
                payload: {
                    income: tmpIncome,
                    totalGoalsAmount: tmpTotalGoalsAmount,
                    disposable: tmpDisposable,
                    budgetID: postResult.data.id
                }
            }
        ];

        let newState = INITIAL_BUDGET_STATE;
        const mockCallback = jest.fn();
        store.dispatch(await budgetActions.createBudget(
            tmpIncome, tmpDisposable, tmpTotalGoalsAmount, mockCallback)).then(() => {
            newState = authReducer(newState, expectedAction[1]);
            expect(store.getState().auth).toEqual(newState);
        });

        const afterWrapper = shallow(
            <CreateBudget/>,
            {context: {store}}
        );
        expect(afterWrapper.dive()).toMatchSnapshot();
    });
});

describe('Accounts', () => {
    beforeEach(() => {
        setupFirebaseMock();
    });

    it('renders SignIn, envoke sign in action, update state and rerender SignIn', async () => {
        // Setup initial screen
        const initialWrapper = shallow(
            <Accounts/>,
            {context: {store}}
        );
        expect(initialWrapper.dive()).toMatchSnapshot();

        // Fire action
        const expectedAction = [
            {type: GET_LINKED_ACCOUNTS},
            {
                type: GET_LINKED_ACCOUNTS_SUCCESS,
                payload: ["TEST_ACCOUNT"]
            }
        ];

        let newState = INITIAL_ACCOUNT_STATE;
        store.dispatch(await accountActions.getLinkedAccounts()).then(() => {
            newState = accountReducer(newState, expectedAction[1]);
            console.log(store.getState());
            expect(store.getActions()).toEqual(expectedAction);
        });

        const afterWrapper = shallow(
            <Accounts/>,
            {context: {store}},
        );
        expect(afterWrapper.dive()).toMatchSnapshot();
    });
});

describe('ExpenseOverview', () => {
    beforeEach(() => {
        setupFirebaseMock();
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    it('renders ExpenseOverview, envoke getExpensesOfMonth action, update state and rerender ExpenseOverview', async () => {
        // Setup initial screen
        const initialWrapper = shallow(
            <ExpenseOverview/>,
            {context: {store}}
        );
        expect(initialWrapper.dive()).toMatchSnapshot();

        // Fire action
        const amountVal = 10;
        const totalAmount = amountVal * 3;
        const returnedExpenses = {
            data: [
                {amount: amountVal},
                {amount: amountVal},
                {amount: amountVal}
            ]
        };
        axios.get.mockResolvedValueOnce(returnedExpenses);

        const expectedAction = [
            {type: GET_EXPENSES_OF_MONTH},
            {
                type: GET_EXPENSES_OF_MONTH_SUCCESS,
                payload: returnedExpenses, totalAmount
            }
        ];

        store.dispatch(await expenseActions.getExpensesOfMonth()).then(() => {
            expect(store.getActions()).toEqual(expectedAction);
        });

        const afterWrapper = shallow(
            <ExpenseOverview/>,
            {context: {store}},
        );
        expect(afterWrapper.dive()).toMatchSnapshot();
    });
});

describe('EditDisposable', () => {
    beforeEach(() => {
        setupFirebaseMock();
    });

    afterEach(() => {
        jest.resetAllMocks();
    });
    it('renders EditDisposable, envoke editDisposable action, update state and rerender EditDisposable', async () => {
        // Setup initial screen
        const initialWrapper = shallow(
            <EditDisposable/>,
            {context: {store}}
        );
        expect(initialWrapper.dive()).toMatchSnapshot();

        // Fire action
        const testBudgetID = "123";
        const tmpDisposable = 456;
        const categoryDisposableItems = [{categoryID: "abc", newAmount: 50}];
        axios.post.mockResolvedValueOnce();

        const expectedAction = [
            {type: EDIT_DISPOSABLE},
            {
                type: EDIT_DISPOSABLE_SUCCESS,
                payload: tmpDisposable
            }
        ];

        store.dispatch(await disposableActions.editDisposable({
            budgetID: testBudgetID, categoryDisposableItems, tmpDisposable})).then(() => {
            expect(store.getActions()).toEqual(expectedAction);
        });

        const afterWrapper = shallow(
            <EditDisposable/>,
            {context: {store}},
        );
        expect(afterWrapper.dive()).toMatchSnapshot();
    });
});