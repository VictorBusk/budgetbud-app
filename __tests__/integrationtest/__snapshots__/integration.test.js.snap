// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integrationtest - Testing SignIn renders as expected 1`] = `
<KeyboardAvoidingView
  behavior="padding"
  keyboardVerticalOffset={0}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#03426A",
      "flex": 2,
      "justifyContent": "center",
    }
  }
>
  <TouchableWithoutFeedback
    accessible={false}
    onPress={[Function]}
  >
    <Styled(Container)
      style={
        Object {
          "alignSelf": "stretch",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "flex-end",
            },
            Object {
              "flex": 2,
            },
          ]
        }
      >
        <Logo
          logoText="Welcome to BudgetBud"
        />
      </View>
      <AuthForm
        cprNumber="1234567890"
        handleSubmit={[Function]}
        isSignIn={true}
        onCprNumberChange={[Function]}
        onSecondInputChange={[Function]}
        secondInput="1234"
      />
      <Styled(Container)>
        <Styled(Container)
          style={
            Object {
              "alignItems": "flex-start",
              "alignSelf": "center",
              "flexShrink": 1,
              "paddingTop": 13,
              "width": "80%",
            }
          }
        >
          <Styled(Button)
            onPress={[Function]}
            style={
              Object {
                "height": 30,
                "paddingLeft": 5,
              }
            }
            transparent={true}
          >
            <Styled(Label)
              style={
                Object {
                  "color": "rgba(255,255,255,0.6)",
                }
              }
            >
              Forgot code?
            </Styled(Label)>
          </Styled(Button)>
          <Styled(Button)
            onPress={[Function]}
            style={
              Object {
                "height": 30,
                "paddingLeft": 5,
              }
            }
            transparent={true}
          >
            <Styled(Label)
              style={
                Object {
                  "color": "rgba(255,255,255,0.6)",
                }
              }
            >
              New user?
            </Styled(Label)>
          </Styled(Button)>
        </Styled(Container)>
      </Styled(Container)>
    </Styled(Container)>
  </TouchableWithoutFeedback>
</KeyboardAvoidingView>
`;
